<div style="border:solid green 2px; padding: 20px">

**Hello Ivey,**

My name is **John Dickson** (https://hub.tripleten.com/u/3cb57352) and today I'll be reviewing your project!

You’ll find specific notes inside the project file, marked green, yellow or red.


**Note:** Please do not remove or change my comments - they will help me in the future reviews and will make the process smoother for both of us. 

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment</b> 
    
Such comment will mark efficient solutions and good ideas that can be used in other projects. It will also point at the document formatting, which was done for you in this project, but you will need to do it yourself in the future ones.
</div>

<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment</b> 
    
The parts marked with yellow comments indicate that there is room for optimisation. Though the correction is not necessary it is good if you implement it.
</div>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment</b> 
    
If you see such a comment, it means that there is a problem that needs to be fixed. Please note that I won't be able to accept your project until the issue is resolved.
</div>

---
    
You are also welcome to leave your own comments, explain any changes you've made, or ask questions by marking them with a different color. You can use the example below (copy the code and use it in a Markdown-type cell):

```
    
<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student’s Comment</b>

```
    
It will appear like this:
    
<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student’s Comment</b></div>
</div>

<div style="border:solid Purple 2px; padding: 20px">

V1

<div style="border: 2px solid Green; padding: 5px; margin: 5px">
 
**What Was Great:**

- Great overall structure to the project
- Outstanding analysis, this is what we like to see in the projects
- Great use of functions throughout the project.

</div>
<div style="border: 2px solid red; padding: 5px; margin: 5px">

**What could be improved:**

- Hard coding the average reserves, it is a bad habbit to hard code values from something that is calculated, because if something changes we then have to manually update all the values.
- Repeatedly defining the same variables. If one of them changes we have to then find all the places they are defined and fix them.
- Retraining the model when we have the predictions and targets already. Try to avoid repeating code. 

</div>

---

Great Project Ivey! Just a couple of small issues to address, such as hard coding values that are calculated, and duplicating models. Clean up those small issues and you will easily move in in the next itteration. 

<div style="border:solid Purple 2px; padding: 20px">

V2

<div style="border: 2px solid Green; padding: 5px; margin: 5px">
 
**What Was Great:**

- Cleaned up the repeat code
- Changed hard coded values to use calculated values
- removed repeat variable definitions. 

</div>


---

Great work Ivey, this project has cleaned up nicely. Very well done, I look forward to seeing future projects from you. 

Conditions For Approval:

Only Linear Regression Is Suitable For Model Training
A Study Of 500 Points Is Carried Wuth Picking The Best 200 Points For The Profit Calculation When Exploring A Region
Budget For Development Of 200 Oil Wells Is 100 USD Million
One Barrel Of Raw Materials Brings 4.5 USD Of Revenue
The Revenue From One Unit Of Product Is 4,500 Dollars (Volume Of Reserve Is In Thousand Barrels)
After Risk Evaluation, Regions With The Risk Of Losses Lower Than 2.5% Are Kept
From The Ones That Fit The Criteria, The Region With The Highest Average Profit Is Selected

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, KFold, cross_val_score
from sklearn.metrics import mean_squared_error, make_scorer

import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

In [2]:
#Download Data
geo0 = pd.read_csv('/datasets/geo_data_0.csv')
geo1 = pd.read_csv('/datasets/geo_data_1.csv')
geo2 = pd.read_csv('/datasets/geo_data_2.csv')

#Read Daata
print(geo0.head(1))
print(geo1.head(1))
print(geo2.head(1))

      id        f0        f1       f2     product
0  txEyH  0.705745 -0.497823  1.22117  105.280062
      id         f0     f1        f2   product
0  kBEdx -15.001348 -8.276 -0.005876  3.179103
      id        f0        f1        f2    product
0  fwXo0 -1.146987  0.963328 -0.828965  27.758673


In [3]:
#Check For Missing Values And Check Data Types
geo0.info()
geo1.info()
geo2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null 

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment V1</b> 
    
Data loaded and checked!
</div>

No Missing Values. Data Types Are Correct. 

In [4]:
# Split, Train And Test The Model For Each Region
def train_and_evaluate_region(data, region_name):
    features = data[['f0', 'f1', 'f2']]
    target = data['product']
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=12345)

    model = LinearRegression()
    model.fit(features_train, target_train)

    preds = model.predict(features_valid)

    #Calculate RMSE And Average Predicted Volume
    rmse = mean_squared_error(target_valid, preds, squared=False)
    avg_pred = preds.mean()
    avg_actual = target_valid.mean()

    print(f'📍 {region_name}')
    print(f'RMSE: {rmse:.3f}')
    print(f'Average Predicted Reserves: {avg_pred:.3f}')
    print(f'Average actual reserves: {avg_actual:.3f}')
    print('-' * 45)

    return model, features_valid, target_valid, preds

#Run Models For Each Region
model_0, fv0, tv0, pred0 = train_and_evaluate_region(geo0, 'Region 0')
model_1, fv1, tv1, pred1 = train_and_evaluate_region(geo1, 'Region 1')
model_2, fv2, tv2, pred2 = train_and_evaluate_region(geo2, 'Region 2')


📍 Region 0
RMSE: 37.579
Average Predicted Reserves: 92.593
Average actual reserves: 92.079
---------------------------------------------
📍 Region 1
RMSE: 0.893
Average Predicted Reserves: 68.729
Average actual reserves: 68.723
---------------------------------------------
📍 Region 2
RMSE: 40.030
Average Predicted Reserves: 94.965
Average actual reserves: 94.884
---------------------------------------------


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment V1</b> 
    
Great!
</div>

### 🔍 Step 2.5 — Model Performance Analysis

After training and testing the linear regression model across all three regions, the following results were obtained:

| Region | RMSE | Avg Predicted Reserves (thousand barrels) | Avg Actual Reserves (thousand barrels) |
|:--|:--|:--|:--|
| **Region 0** | **37.579** | 92.593 | 92.079 |
| **Region 1** | **0.893** | 68.729 | 68.723 |
| **Region 2** | **40.030** | 94.965 | 94.884 |

#### **Interpretation:**

- **Region 0:**  
  Shows moderate predictive accuracy. Although average reserves are high, the RMSE (37.579) suggests the model struggles to make stable predictions, introducing uncertainty in expected profit calculations.

- **Region 1:**  
  Exhibits *excellent* predictive performance. The RMSE (0.893) is nearly zero, and the predicted and actual average reserves are almost identical. This means the model fits this region's data extremely well and can be trusted for profit forecasting.

- **Region 2:**  
  Has the highest average predicted reserves, but with an RMSE of 40.030 — indicating that while reserves are large, predictions vary significantly. This makes the region riskier compared to Region 1.

#### **Conclusion:**
Region 1 currently stands out as the most reliable and predictable location based on model performance.  
However, the final decision will depend not only on accuracy but also on **profitability and risk**, which will be analyzed next using the **Bootstrapping technique**.

In [5]:
#Prepare for Profit Calculation
#Business Considerations
Budget = 100_000_000 # $100 Million Budget
top_wells = 200 #Develop 200 Wells
price_per_barrel = 4.5 #$ Per Barrel
barrel_unit = 1000 #1 unit equals 1000 barrels
price_per_unit = price_per_barrel * barrel_unit
Study_size = 500
Random_state = 12345
sample_size = 1000

#Average Reserves
avg_reserves = {
    'Region 0': pred0.mean(),
    'Region 1': pred1.mean(),
    'Region 2': pred2.mean()
}

#Calculate Minimum Volume Of Reserves Needed To Break Even
min_volume_needed = Budget / (top_wells * price_per_barrel * barrel_unit)
print(f'Minimum Volume Of Reserves Needed For Profit: {min_volume_needed:.2f} Thousand Barrels')

#Compare To Average Reserves Per Region
for region, avg in avg_reserves.items():
    status = '✅ Above Threshold' if avg > min_volume_needed else '⚠️ Below Threshold'
    print(f'{region}: {avg:.2f} → {status}')


Minimum Volume Of Reserves Needed For Profit: 111.11 Thousand Barrels
Region 0: 92.59 → ⚠️ Below Threshold
Region 1: 68.73 → ⚠️ Below Threshold
Region 2: 94.97 → ⚠️ Below Threshold


<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment V1</b> 
    
We should not be hard coding the average reserves here, we have them calcuated above, how about we pass them out in variables and use that? 
</div>

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment V2</b> 
    
Great! That is what we want to see. 
</div>

### 💰 Step 3 — Profitability Threshold Analysis

After applying the business conditions, the **minimum volume of reserves required for a profitable development** was calculated as:

> **111.11 thousand barrels per well**

| Region | Average Predicted Reserves (thousand barrels) | Profitability Status |
|:--|:--|:--|
| **Region 0** | 92.59 | ⚠️ Below Threshold |
| **Region 1** | 68.73 | ⚠️ Below Threshold |
| **Region 2** | 94.97 | ⚠️ Below Threshold |

#### **Interpretation:**
- The average predicted reserves in **all three regions** are **below the break-even point** of 111.11 thousand barrels.  
- This means that **none of the regions would be profitable** if development costs were applied to the full dataset without selection.  
- However, this doesn’t mean the project is unprofitable overall — in the next stage, we’ll **simulate selecting only the most promising wells** to optimize returns and **analyze potential profit and risk using Bootstrapping**.

#### **Next Step:**
We’ll proceed to **Step 4**, where we:
1. Select the top 200 wells per region based on model predictions.  
2. Calculate total profit from those wells.  
3. Use the Bootstrapping technique to measure **average profit, risk of loss, and 95% confidence intervals** for each region.

In [6]:
# Profit from model predictions

#Create Function To Predict Profits From Regions
def profit_from_predictions(target_valid, target_pred, n_select=top_wells, price=price_per_unit, budget=Budget):
    target_pred_s = pd.Series(target_pred, index=target_valid.index)
    top_idx = target_pred_s.sort_values(ascending=False).index[:n_select]

    total_selected_reserves = target_valid.loc[top_idx].sum()
    revenue = price * total_selected_reserves
    profit = revenue - budget

    return profit, revenue, total_selected_reserves, top_idx

In [7]:
def eval_region_revenue(model, name, target_valid, preds):

    profit, revenue, total_res, top_idx = profit_from_predictions(target_valid, preds)

    avg_pred = pd.Series(preds, index=target_valid.index).loc[top_idx].mean()
    avg_actual = target_valid.loc[top_idx].mean()

    print(f'📍 {name}')
    print(f'Selected Wells: {len(top_idx)}')
    print(f'Total Reserves (Actual) Of Selected Wells: {total_res:,.2f} Thousand Barrels')
    print(f'Revenue (@ $4,500 Per Unit): ${revenue:,.0f}')
    print(f'Profit After $100M Budget: ${profit:.0f}')
    print(f'Avg Predicted (Selected): {avg_pred:.2f} | Avg Actual (Selected): {avg_actual:.2f}')
    print('-'*60)

eval_region_revenue(model_0, 'Region 0', tv0, pred0)
eval_region_revenue(model_1, 'Region 1', tv1, pred1)
eval_region_revenue(model_2, 'Region 2', tv2, pred2)

📍 Region 0
Selected Wells: 200
Total Reserves (Actual) Of Selected Wells: 29,601.84 Thousand Barrels
Revenue (@ $4,500 Per Unit): $133,208,260
Profit After $100M Budget: $33208260
Avg Predicted (Selected): 155.51 | Avg Actual (Selected): 148.01
------------------------------------------------------------
📍 Region 1
Selected Wells: 200
Total Reserves (Actual) Of Selected Wells: 27,589.08 Thousand Barrels
Revenue (@ $4,500 Per Unit): $124,150,867
Profit After $100M Budget: $24150867
Avg Predicted (Selected): 138.73 | Avg Actual (Selected): 137.95
------------------------------------------------------------
📍 Region 2
Selected Wells: 200
Total Reserves (Actual) Of Selected Wells: 28,245.22 Thousand Barrels
Revenue (@ $4,500 Per Unit): $127,103,500
Profit After $100M Budget: $27103500
Avg Predicted (Selected): 148.02 | Avg Actual (Selected): 141.23
------------------------------------------------------------


### 💵 Step 4 — Profit Calculation for Selected Wells

After training the models, we selected the **top 200 wells** in each region based on predicted reserves and computed profit using:

- **Revenue per thousand barrels:** \$4,500  
- **Development budget:** \$100 million  
- **Selection size:** 200 wells per region  

---

#### 📍 Region 0
- **Total Actual Reserves:** 29,601.84 thousand barrels  
- **Revenue:** \$133,208,260  
- **Profit:** **\$33,208,260**  
- **Avg Predicted (Selected):** 155.51  
- **Avg Actual (Selected):** 148.01  

---

#### 📍 Region 1
- **Total Actual Reserves:** 27,589.08 thousand barrels  
- **Revenue:** \$124,150,867  
- **Profit:** **\$24,150,867**  
- **Avg Predicted (Selected):** 138.73  
- **Avg Actual (Selected):** 137.95  

---

#### 📍 Region 2
- **Total Actual Reserves:** 28,245.22 thousand barrels  
- **Revenue:** \$127,103,500  
- **Profit:** **\$27,103,500**  
- **Avg Predicted (Selected):** 148.02  
- **Avg Actual (Selected):** 141.23  

---

### 🔍 Interpretation

- All regions yield **positive profit** under the given conditions.  
- **Region 0** delivers the **highest total profit (\$33.2M)**, suggesting stronger revenue potential despite its higher RMSE in earlier tests.  
- **Region 1**, while extremely accurate, generates the lowest profit due to smaller overall reserves.  
- **Region 2** performs moderately, balancing accuracy and output volume.  

---

### 🧭 Conclusion

While **Region 0** appears most profitable in this initial simulation, **risk must still be assessed**.  
Next, we’ll apply the **Bootstrapping technique** (Step 5) to simulate variability, compute **confidence intervals**, and estimate **risk of losses** for each region before making the final investment recommendation.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment V1</b> 
    
Great analysis!
</div>

In [8]:
#Bootstrapping profit, CI, and risk of loss

#Create Validation Dataframe With Preds & Actuals
def create_valid_df(model, features_valid, target_valid, preds):
    valid_df = features_valid.copy()
    valid_df['pred'] = preds
    valid_df['actual'] = target_valid.values
    return valid_df

valid_df0 = create_valid_df(model_0, fv0, tv0, pred0)
valid_df1 = create_valid_df(model_1, fv1, tv1, pred1)
valid_df2 = create_valid_df(model_2, fv2, tv2, pred2)

#Choose Top 200 Wells By Predicted, Compute Profit Using Actual Reserves
def profit_from_sample(sample_df):
    top = sample_df.sort_values('pred', ascending=False).head(top_wells)
    revenue = top['actual'].sum() * price_per_unit
    profit = revenue - Budget
    return profit

def bootstrap_profit(valid_df, n_boot=sample_size, study_size=Study_size, seed=Random_state, return_samples=False):
    rng = np.random.default_rng(seed)
    profits = np.empty(n_boot, dtype=float)
    n = len(valid_df)
    # If Validation Set < Study Size, Sample With Replacement Anyway
    for b in range(n_boot):
        idx = rng.integers(0, n, size=study_size)
        sample = valid_df.iloc[idx]
        profits[b] = profit_from_sample(sample)
    mean_profit = profits.mean()
    ci_low, ci_high = np.percentile(profits, [2.5, 97.5])
    risk_of_loss = (profits < 0).mean() #Probability Of Negative Profit
    out = {
        'mean_profit': mean_profit,
        'ci_low': ci_low,
        'ci_high': ci_high,
        'risk_of_loss': risk_of_loss,
        'profits': profits
    }
    if return_samples:
        out['profits'] = profits
    return out
def run_region(valid_df, name):
    stats = bootstrap_profit(valid_df, n_boot=sample_size, study_size=Study_size, seed=Random_state)
    print(f'📍 {name}')
    print(f'Mean Profit: ${stats["mean_profit"]:,.0f}')
    print(f'95% CI: [${stats["ci_low"]:,.0f}, ${stats["ci_high"]:,.0f}]')
    print(f'Risk Of Loss (<$0): {stats["risk_of_loss"]*100:.2f}%')
    print('-'*60)
    return stats

#Run All Three Regions (geo0, geo1, geo2)
stats0 = run_region(valid_df0, 'Region 0')
stats1 = run_region(valid_df1, 'Region 1')
stats2 = run_region(valid_df2, 'Region 2')

#Stop Samples From Printing
pretty0 = {k: v for k, v in stats0.items() if k != 'profits'}
pretty1 = {k: v for k, v in stats1.items() if k != 'profits'}
pretty2 = {k: v for k, v in stats2.items() if k != 'profits'}

print(pretty0)
print(pretty1)
print(pretty2)

📍 Region 0
Mean Profit: $3,690,527
95% CI: [$-1,767,205, $8,869,696]
Risk Of Loss (<$0): 8.50%
------------------------------------------------------------
📍 Region 1
Mean Profit: $4,541,582
95% CI: [$488,527, $8,459,902]
Risk Of Loss (<$0): 1.40%
------------------------------------------------------------
📍 Region 2
Mean Profit: $3,788,329
95% CI: [$-1,350,656, $8,823,111]
Risk Of Loss (<$0): 6.00%
------------------------------------------------------------
{'mean_profit': 3690526.7292691194, 'ci_low': -1767205.238772129, 'ci_high': 8869696.166351613, 'risk_of_loss': 0.085}
{'mean_profit': 4541582.116482314, 'ci_low': 488526.54075439094, 'ci_high': 8459901.657637987, 'risk_of_loss': 0.014}
{'mean_profit': 3788328.7285128594, 'ci_low': -1350655.6768556158, 'ci_high': 8823111.12098347, 'risk_of_loss': 0.06}


<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment</b> 
    
<strike>There is a lot of repetition for things you have already done here. We already have predictions and targets from when we trained our models earlier. 

<strike>You have defined several of your variables such as price and budget before, you don't want to have that in several places. 

</div>

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment V1</b> 
    
Great bootstrapping and profit calculation, the results appear to be what is expected.
</div>

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment V2</b> 
    
Great! That is cleaned up nicely.
</div>


## 🧭 Step 5 — Risk Analysis (Bootstrapping)

Using bootstrapping (1,000 samples, study size = 500), we evaluated **profit variability**, **confidence intervals**, and **risk of loss** for each region. This analysis allows us to assess not only which region performs best on average but also which maintains stability under uncertainty.

---

### 🔍 Analysis & Interpretation

**Region 1** emerged as the **most stable and reliable performer**.  
Its mean profit (~$4.5M) was accompanied by the **lowest risk of loss (1.4%)**, with confidence intervals that remained comfortably positive across most samples.  
This consistency indicates a **highly predictable revenue stream** and strong resilience against negative variance — aligning well with the company’s acceptable risk threshold (<2.5%).

**Region 0**, although showing the **highest potential upside (~$8.8M)**, also exhibited a **greater volatility**, reflected by its 8.5% chance of loss.  
This means while Region 0 could yield larger profits in favorable conditions, the probability of financial underperformance is significantly higher, making it a **riskier investment option**.

**Region 2** demonstrated **moderate profitability and risk**, averaging around $3.8M profit with a 6% risk of loss.  
It represents a balanced but less competitive outcome when compared to the stability and efficiency of Region 1.

---

### 🧾 Conclusion

While Region 0 achieved the highest profit potential in earlier simulations, **bootstrapping revealed that Region 1 delivers the most dependable return** under variable market conditions.  
Its low variance, minimal downside exposure, and consistent profitability make **Region 1 the optimal investment choice** for the company’s current objectives.

✅ **Final Recommendation:**  
**Invest in Region 1** — it provides the best balance between profitability and risk, ensuring long-term financial sustainability.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment V1</b> 
    
Great summery and analysis.
</div>